File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/declaration/ComponentDeclaration.java
package ru.tinkoff.kora.kora.app.annotation.processor.declaration;

import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;

import javax.lang.model.element.*;
import javax.lang.model.type.TypeMirror;
import java.util.List;
import java.util.Set;

public sealed interface ComponentDeclaration {
    TypeMirror type();

    Element source();

    Set<String> tags();

    default boolean isTemplate() {
        return TypeParameterUtils.hasTypeParameter(this.type());
    }

    default boolean isDefault() {
        return false;
    }

    record FromModuleComponent(TypeMirror type, ModuleDeclaration module, Set<String> tags, ExecutableElement method, List<TypeMirror> methodParameterTypes,
                               List<TypeMirror> typeVariables) implements ComponentDeclaration {
        @Override
        public Element source() {
            return this.method;
        }

        @Override
        public boolean isDefault() {
            return AnnotationUtils.findAnnotation(this.method, CommonClassNames.defaultComponent) != null;
        }
    }

    record AnnotatedComponent(TypeMirror type, TypeElement typeElement, Set<String> tags, ExecutableElement constructor, List<TypeMirror> methodParameterTypes,
                              List<TypeMirror> typeVariables) implements ComponentDeclaration {
        @Override
        public Element source() {
            return this.constructor;
        }
    }

    record DiscoveredAsDependencyComponent(TypeMirror type, TypeElement typeElement, ExecutableElement constructor) implements ComponentDeclaration {
        public DiscoveredAsDependencyComponent {
            assert typeElement.getTypeParameters().isEmpty();
        }

        @Override
        public Element source() {
            return this.constructor;
        }

        @Override
        public Set<String> tags() {
            return Set.of();
        }

        @Override
        public boolean isTemplate() {
            return false;
        }
    }

    record FromExtensionComponent(TypeMirror type, ExecutableElement sourceMethod, List<TypeMirror> methodParameterTypes) implements ComponentDeclaration {
        @Override
        public Element source() {
            return this.sourceMethod;
        }

        @Override
        public Set<String> tags() {
            return Set.of();
        }
    }

    record PromisedProxyComponent(TypeElement typeElement, TypeMirror type, com.squareup.javapoet.ClassName className) implements ComponentDeclaration {
        public PromisedProxyComponent(TypeElement typeElement, com.squareup.javapoet.ClassName className) {
            this(typeElement, typeElement.asType(), className);
        }

        public PromisedProxyComponent withType(TypeMirror type) {
            return new PromisedProxyComponent(typeElement, type, className);
        }


        @Override
        public Element source() {
            return this.typeElement;
        }

        @Override
        public Set<String> tags() {
            return Set.of(CommonClassNames.promisedProxy.canonicalName());
        }
    }

    record OptionalComponent(TypeMirror type, Set<String> tags) implements ComponentDeclaration {
        @Override
        public Element source() {
            return null;
        }
    }

    static ComponentDeclaration fromModule(ModuleDeclaration module, ExecutableElement method) {
        var type = method.getReturnType();
        var tags = TagUtils.parseTagValue(method);
        var parameterTypes = method.getParameters().stream().map(VariableElement::asType).toList();
        var typeParameters = method.getTypeParameters().stream().map(TypeParameterElement::asType).toList();
        return new FromModuleComponent(type, module, tags, method, parameterTypes, typeParameters);
    }

    static ComponentDeclaration fromAnnotated(TypeElement typeElement) {
        var constructors = CommonUtils.findConstructors(typeElement, m -> m.contains(Modifier.PUBLIC));
        if (constructors.size() != 1) {
            throw new ProcessingErrorException("@Component annotated class should have exactly one public constructor", typeElement);
        }
        var constructor = constructors.get(0);
        var type = typeElement.asType();
        var tags = TagUtils.parseTagValue(typeElement);
        var parameterTypes = constructor.getParameters().stream().map(VariableElement::asType).toList();
        var typeParameters = typeElement.getTypeParameters().stream().map(TypeParameterElement::asType).toList();
        return new AnnotatedComponent(type, typeElement, tags, constructor, parameterTypes, typeParameters);
    }

    static ComponentDeclaration fromDependency(TypeElement typeElement) {
        var constructors = CommonUtils.findConstructors(typeElement, m -> m.contains(Modifier.PUBLIC));
        if (constructors.size() != 1) {
            throw new ProcessingErrorException("Can't create component from discovered as dependency class: class should have exactly one public constructor", typeElement);
        }
        var constructor = constructors.get(0);
        var type = typeElement.asType();
        var tags = TagUtils.parseTagValue(typeElement);
        if (!tags.isEmpty()) {
            throw new ProcessingErrorException("Discovered as dependency class cannot have tags", typeElement);
        }
        return new DiscoveredAsDependencyComponent(type, typeElement, constructor);
    }

    static ComponentDeclaration fromExtension(ExtensionResult.GeneratedResult generatedResult) {
        var sourceMethod = generatedResult.sourceElement();
        var parameterTypes = sourceMethod.getParameters().stream().map(VariableElement::asType).toList();
        if (sourceMethod.getKind() == ElementKind.CONSTRUCTOR) {
            var typeElement = (TypeElement) sourceMethod.getEnclosingElement();
            var type = typeElement.asType();
            return new FromExtensionComponent(type, sourceMethod, parameterTypes);
        } else {
            var type = sourceMethod.getReturnType();
            return new FromExtensionComponent(type, sourceMethod, parameterTypes);
        }
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/declaration/ComponentDeclaration.kt
package ru.tinkoff.kora.kora.app.ksp.declaration

import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.TypeName
import ru.tinkoff.kora.kora.app.ksp.ProcessingContext
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.KspCommonUtils.fixPlatformType
import ru.tinkoff.kora.ksp.common.TagUtils
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException

sealed interface ComponentDeclaration {
    val type: KSType
    val source: KSDeclaration
    val tags: Set<String>

    fun isTemplate(): Boolean {
        for (argument in type.arguments) {
            if (argument.hasGenericVariable()) {
                return true
            }
        }
        return false
    }

    fun isDefault(): Boolean {
        return false
    }

    data class FromModuleComponent(
        override val type: KSType,
        val module: ModuleDeclaration,
        override val tags: Set<String>,
        val method: KSFunctionDeclaration,
        val methodParameterTypes: List<KSType>,
        val typeVariables: List<KSTypeArgument>
    ) : ComponentDeclaration {
        override val source get() = this.method
        override fun isDefault(): Boolean {
            return method.findAnnotation(CommonClassNames.defaultComponent) != null
        }
    }

    data class AnnotatedComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        override val tags: Set<String>,
        val constructor: KSFunctionDeclaration,
        val methodParameterTypes: List<KSType>,
        val typeVariables: List<KSTypeArgument>
    ) : ComponentDeclaration {
        override val source get() = this.constructor
    }

    data class DiscoveredAsDependencyComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        val constructor: KSFunctionDeclaration,
        override val tags: Set<String>
    ) : ComponentDeclaration {
        override val source get() = this.constructor
    }

    data class FromExtensionComponent(
        override val type: KSType,
        val sourceMethod: KSFunctionDeclaration,
        val methodParameterTypes: List<KSType>,

        ) : ComponentDeclaration {
        override val source get() = this.sourceMethod
        override val tags get() = setOf<String>()

    }

    data class PromisedProxyComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        val className: TypeName

    ) : ComponentDeclaration {
        override val source get() = this.classDeclaration
        override val tags get() = setOf(CommonClassNames.promisedProxy.canonicalName)
    }


    data class OptionalComponent(
        override val type: KSType,
        override val tags: Set<String>
    ) : ComponentDeclaration {
        override val source get() = type.declaration
    }


    companion object {
        fun fromModule(ctx: ProcessingContext, module: ModuleDeclaration, method: KSFunctionDeclaration): FromModuleComponent {
            // modules can be written in java so we better fix platform nullability
            val type = method.returnType!!.resolve().fixPlatformType(ctx.resolver)
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", method)
            }
            val tags = TagUtils.parseTagValue(method)
            val parameterTypes = method.parameters.map { it.type.resolve().fixPlatformType(ctx.resolver) }
            val typeParameters = method.typeParameters.map {
                val t = it.bounds.firstOrNull()?.resolve()?.fixPlatformType(ctx.resolver) ?: ctx.resolver.builtIns.anyType

                ctx.resolver.getTypeArgument(
                    ctx.resolver.createKSTypeReferenceFromKSType(t),
                    it.variance
                )
            }
            return FromModuleComponent(type, module, tags, method, parameterTypes, typeParameters)
        }

        fun fromAnnotated(ctx: ProcessingContext, classDeclaration: KSClassDeclaration): AnnotatedComponent {
            val constructor = classDeclaration.primaryConstructor
            if (constructor == null) {
                throw ProcessingErrorException("@Component annotated class should have primary constructor", classDeclaration)
            }
            val typeParameters = classDeclaration.typeParameters.map {
                val t = it.bounds.firstOrNull()?.resolve() ?: ctx.resolver.builtIns.anyType

                ctx.resolver.getTypeArgument(
                    ctx.resolver.createKSTypeReferenceFromKSType(t),
                    it.variance
                )
            }
            val type = classDeclaration.asType(classDeclaration.typeParameters.map { ctx.resolver.getTypeArgument(it.bounds.first(), it.variance) })
            val tags = TagUtils.parseTagValue(classDeclaration)
            val parameterTypes = constructor.parameters.map { it.type.resolve() }

            return AnnotatedComponent(type, classDeclaration, tags, constructor, parameterTypes, typeParameters)
        }

        fun fromDependency(ctx: ProcessingContext, classDeclaration: KSClassDeclaration): DiscoveredAsDependencyComponent {
            val constructor = classDeclaration.primaryConstructor
            if (constructor == null) {
                throw ProcessingErrorException("No primary constructor to parse component for: $classDeclaration", classDeclaration)
            }
            val type = classDeclaration.asType(listOf())
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", classDeclaration)
            }
            val tags = TagUtils.parseTagValue(classDeclaration)

            return DiscoveredAsDependencyComponent(type, classDeclaration, constructor, tags)
        }

        fun fromExtension(ctx: ProcessingContext, extensionResult: ExtensionResult.GeneratedResult): FromExtensionComponent {
            val sourceMethod = extensionResult.constructor
            val sourceType = extensionResult.type
            val parameterTypes = sourceType.parameterTypes.map { it!!.fixPlatformType(ctx.resolver) }
            val type = sourceType.returnType!!
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", sourceMethod)
            }
            return FromExtensionComponent(type, sourceMethod, parameterTypes)
        }
    }
}

private fun KSTypeArgument.hasGenericVariable(): Boolean {
    val type = this.type
    if (type == null) {
        return true
    }
    val resolvedType = type.resolve()
    if (resolvedType.declaration is KSTypeParameter) {
        return true
    }

    for (param in resolvedType.arguments) {
        if (param.hasGenericVariable()) {
            return true
        }
    }
    return false
}
